File: backend/models/hospital.py
"""Hospital facility, department, bed, encounter, and admission ORM models."""
from datetime import datetime, timezone

from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from ..database import Base


class HospitalFacility(Base):
    __tablename__ = "hospital_facilities"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, unique=True, index=True)
    facility_type = Column(String, default="hospital")
    country = Column(String, nullable=True)
    region = Column(String, nullable=True)
    status = Column(String, default="active")
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))


class Department(Base):
    __tablename__ = "departments"

    id = Column(Integer, primary_key=True, index=True)
    facility_id = Column(Integer, ForeignKey("hospital_facilities.id"), nullable=True, index=True)
    name = Column(String, unique=True, index=True)
    department_type = Column(String)  # OPD, IPD, Emergency, Diagnostics, Pharmacy
    location = Column(String, nullable=True)
    description = Column(Text, nullable=True)
    status = Column(String, default="active")
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))

    facility = relationship("HospitalFacility")


class Bed(Base):
    __tablename__ = "beds"

    id = Column(Integer, primary_key=True, index=True)
    facility_id = Column(Integer, ForeignKey("hospital_facilities.id"), nullable=True, index=True)
    department_id = Column(Integer, ForeignKey("departments.id"))
    bed_number = Column(String, index=True)
    ward = Column(String, nullable=True)
    status = Column(String, default="available")  # available, occupied, maintenance
    current_patient_id = Column(Integer, ForeignKey("users.id"), nullable=True, index=True)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))

    facility = relationship("HospitalFacility")
    department = relationship("Department")
    current_patient = relationship("User", foreign_keys=[current_patient_id])


class Encounter(Base):
    __tablename__ = "encounters"

    id = Column(Integer, primary_key=True, index=True)
    facility_id = Column(Integer, ForeignKey("hospital_facilities.id"), nullable=True, index=True)
    patient_id = Column(Integer, ForeignKey("users.id"), index=True)
    doctor_id = Column(Integer, ForeignKey("users.id"), nullable=True, index=True)
    department_id = Column(Integer, ForeignKey("departments.id"), nullable=True)
    encounter_type = Column(String)  # OPD, IPD, Emergency
    reason = Column(Text, nullable=True)
    priority = Column(String, default="routine")  # routine, urgent, emergency
    status = Column(String, default="open")  # open, closed, cancelled
    started_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    ended_at = Column(DateTime, nullable=True)

    facility = relationship("HospitalFacility")
    patient = relationship("User", foreign_keys=[patient_id])
    doctor = relationship("User", foreign_keys=[doctor_id])
    department = relationship("Department")


class Admission(Base):
    __tablename__ = "admissions"

    id = Column(Integer, primary_key=True, index=True)
    facility_id = Column(Integer, ForeignKey("hospital_facilities.id"), nullable=True, index=True)
    encounter_id = Column(Integer, ForeignKey("encounters.id"), index=True)
    patient_id = Column(Integer, ForeignKey("users.id"), index=True)
    doctor_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    department_id = Column(Integer, ForeignKey("departments.id"), nullable=True)
    bed_id = Column(Integer, ForeignKey("beds.id"), nullable=True)
    admitted_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    discharged_at = Column(DateTime, nullable=True)
    reason = Column(Text, nullable=True)
    status = Column(String, default="active")

    facility = relationship("HospitalFacility")
    encounter = relationship("Encounter")
    patient = relationship("User", foreign_keys=[patient_id])
    doctor = relationship("User", foreign_keys=[doctor_id])
    department = relationship("Department")
    bed = relationship("Bed")

File: rust_gateway/src/models/hospital.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

/// Hospital facility tenant entity.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct HospitalFacility {
    pub id: i64,
    pub name: String,
    pub facility_type: String,
    pub country: Option<String>,
    pub region: Option<String>,
    pub status: String,
    pub created_at: Option<NaiveDateTime>,
}

/// Hospital clinical department.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Department {
    pub id: i64,
    pub facility_id: Option<i64>,
    pub name: String,
    pub department_type: String,
    pub location: Option<String>,
    pub description: Option<String>,
    pub status: String,
    pub created_at: Option<NaiveDateTime>,
}

/// Ward inpatient bed.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Bed {
    pub id: i64,
    pub facility_id: Option<i64>,
    pub department_id: i64,
    pub bed_number: String,
    pub ward: Option<String>,
    pub status: String,
    pub current_patient_id: Option<i64>,
    pub created_at: Option<NaiveDateTime>,
}

/// Clinical patient encounter (OPD, IPD, Emergency).
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Encounter {
    pub id: i64,
    pub facility_id: Option<i64>,
    pub patient_id: i64,
    pub doctor_id: Option<i64>,
    pub department_id: Option<i64>,
    pub encounter_type: String,
    pub reason: Option<String>,
    pub priority: String,
    pub status: String,
    pub started_at: Option<NaiveDateTime>,
    pub ended_at: Option<NaiveDateTime>,
    pub is_deleted: i64,
    pub deleted_at: Option<NaiveDateTime>,
}

/// Inpatient hospital admission.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Admission {
    pub id: i64,
    pub facility_id: Option<i64>,
    pub encounter_id: i64,
    pub patient_id: i64,
    pub doctor_id: Option<i64>,
    pub department_id: Option<i64>,
    pub bed_id: Option<i64>,
    pub admitted_at: Option<NaiveDateTime>,
    pub discharged_at: Option<NaiveDateTime>,
    pub reason: Option<String>,
    pub status: String,
    pub is_deleted: i64,
    pub deleted_at: Option<NaiveDateTime>,
}

/// PACS DICOM medical imaging study metadata.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct DicomStudy {
    pub id: i64,
    pub study_uid: String,
    pub patient_id: Option<i64>,
    pub modality: String,
    pub target_vault: String,
    pub file_name: String,
    pub file_size_kb: i64,
    pub is_preamble_valid: String,
    pub created_at: Option<NaiveDateTime>,
}
